python/lib/core/dmod/core/context.py

```diff
@@ -0,0 +1,607 @@
+"""
+Defines a custom Context Manager
+"""
+from __future__ import annotations
+
+import logging
+import multiprocessing
+import os
+import sys
+import threading
+import typing
+import inspect
+import platform
+
+from multiprocessing import managers
+from multiprocessing import RLock
+from multiprocessing import util
+from multiprocessing.context import BaseContext
+from traceback import format_exc
+
+from .decorators import version_range
+
+_PREPARATION_LOCK: RLock = RLock()
+
+SENTINEL = object()
+"""A basic sentinel value to serve as a true 'null' value"""
+
+T = typing.TypeVar("T")
+"""Some generic type of object"""
+
+Manager = typing.TypeVar("Manager", bound=managers.BaseManager, covariant=True)
+"""Any type of manager object"""
+
+ManagerType = typing.Type[Manager]
+"""The type of a manager object itself"""
+
+TypeOfRemoteObject = typing.Union[typing.Type[managers.BaseProxy], type]
+"""A wrapper object that is used to communicate to objects created by Managers"""
+
+_PROXY_TYPE_CACHE: typing.MutableMapping[typing.Tuple[str, typing.Tuple[str, ...]], TypeOfRemoteObject] = {}
+"""A simple mapping of recently created proxies to remote objects"""
+
+__ACCEPTABLE_DUNDERS = (
+    "__getitem__",
+    "__setitem__",
+    "__delitem__",
+    "__contains__",
+    "__call__",
+    "__iter__",
+    "__gt__",
+    "__ge__",
+    "__lt__",
+    "__le__",
+    "__eq__",
+    "__mul__",
+    "__truediv__",
+    "__floordiv__",
+    "__mod__",
+    "__sub__",
+    "__add__",
+    "__ne__",
+    "__get_property__",
+    "__set_property__",
+    "__del_property__"
+)
+"""A collection of dunder names that are valid names for functions on proxies to shared objects"""
+
+PROXY_SUFFIX: typing.Final[str] = "Proxy"
+"""
+Suffix for how proxies are to be named - naming proxies programmatically will ensure they are correctly referenced later
+"""
+
+
+@typing.runtime_checkable
+class ProxiableGetPropertyProtocol(typing.Protocol):
+    """
+    Outline for a class that can explicitly retrieve a property value
+    """
+    def __get_property__(self, key: str) -> typing.Any:
+        ...
+
+@typing.runtime_checkable
+class ProxiableSetPropertyProtocol(typing.Protocol):
+    """
+    Outline for a class that can explicitly set a property value
+    """
+    def __set_property__(self, key: str, value) -> None:
+        ...
+
+
+@typing.runtime_checkable
+class ProxiableDeletePropertyProtocol(typing.Protocol):
+    """
+    Outline for a class that can explicitly delete a property value
+    """
+    def __del_property__(self, key: str) -> None:
+        ...
+
+
+class ProxiablePropertyMixin(ProxiableGetPropertyProtocol, ProxiableSetPropertyProtocol, ProxiableDeletePropertyProtocol):
+    """
+    Mixin functions that allow property functions (fget, fset, fdel) to be called explicitly rather than implicitly
```

Comment:
Ive not finished looking through all the code, so this might be a, "hey, look over there" comment. When and why would you need to call a descriptor explicitly?

Reply:
When descriptors are referenced directly on instances, they of course pretend to be like raw values, not the functions they really are. If you want to forward the value of property `b` in a proxy, you don't forward the property, you instead forward the direct value. This messes up in a lot of places because a lot of the core vanilla python multiprocessing code _only_ deals with callables. _If_ you want to forward a call to a property, you can't do so to the instance member as that returns the value _before_ it may be manipulated. As a result, you have to go from the top down, i.e. reach into the instances _class_, gain the reference to the descriptor for the property, pass in the instance that owns the descriptor, and then get direct access to the functions that may pass the value back and forth along the wire. If you want to pass along the `fget` function and not the value that `fget` produces, you _have_ to pass along `fget`. 

The situation surrounding it is like having a step in a staircase that's half an inch taller than the rest. It doesn't seem like it'll cause a lot of issues, but it ends up causing a **_lot_** of problems if not accounted for.